src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs

```diff
@@ -0,0 +1,299 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal class WinHttpAuthHelper
+    {
+        // TODO:  This looks messy but it is fast. Research a cleaner way
+        // to do this which keeps high performance lookup.
+        //
+        // Fast lookup table to convert WINHTTP_AUTH constants to strings.
+        // WINHTTP_AUTH_SCHEME_BASIC = 0x00000001;
+        // WINHTTP_AUTH_SCHEME_DIGEST = 0x00000008;
+        // WINHTTP_AUTH_SCHEME_NEGOTIATE = 0x00000010;
+        private static readonly string[] s_authSchemeStringMapping =
+        {
+            null,
+            "Basic",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Digest",
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            null,
+            "Negotiate"
+        };
+
+        private static readonly uint[] s_authSchemePriorityOrder =
+        {
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_NEGOTIATE,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_DIGEST,
+            Interop.WinHttp.WINHTTP_AUTH_SCHEME_BASIC
+        };
+
+        // TODO: This current design uses a handler-wide lock to Add/Retrieve
+        // from the cache.  Need to improve this for next iteration in order
+        // to boost performance and scalability.
+        private CredentialCache _credentialCache = new CredentialCache();
+        private object _credentialCacheLock = new object();
```

Comment:
I will make them both `readonly`.

I didn't make them as properties because I was worried about the overhead of using getter's instead of just accessing the field itself. Is there an overhead to worry about or will the compiler be smart and inline?